# Copyright (C) 2018 Sur Herrera Paredes

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

PATRICDATA = '/home/sur/micropopgen/exp/2018/today3/patric'
OUTDIR = '/home/sur/micropopgen/exp/2018/today3/roary'
BINDIR = '/home/sur/micropopgen/src/micropopgen/patric'
specs = ['Streptococcus_anginosus', 'Streptococcus_anginosus']
ids = ['1000570.3', '1000570.8']

rule all:
    input:
        OUTDIR + "/preprocess/Streptococcus_anginosus/1000570.3/1000570.3.gff",
        OUTDIR + "/preprocess/Streptococcus_anginosus/1000570.8/1000570.8.gff",
        OUTDIR + "/preprocess/Streptococcus_anginosus/1000570.3/1000570.3.fna",
        OUTDIR + "/preprocess/Streptococcus_anginosus/1000570.8/1000570.8.fna",
        OUTDIR + "/roary_input/Streptococcus_anginosus-1000570.3.gff",
        OUTDIR + "/roary_input/Streptococcus_anginosus-1000570.8.gff",
        OUTDIR + "/roary/",
        OUTDIR + "/roary/accessory_binary_genes.fa",
        OUTDIR + "/roary/accessory_graph.dot",
        OUTDIR + "/roary/accessory.header.embl",
        OUTDIR + "/roary/accessory.tab",
        OUTDIR + "/roary/blast_identity_frequency.Rtab",
        OUTDIR + "/roary/clustered_proteins",
        OUTDIR + "/roary/core_accessory_graph.dot",
        OUTDIR + "/roary/core_accessory.header.embl",
        OUTDIR + "/roary/core_accessory.tab",
        OUTDIR + "/roary/gene_presence_absence.csv",
        OUTDIR + "/roary/gene_presence_absence.Rtab",
        OUTDIR + "/roary/number_of_conserved_genes.Rtab",
        OUTDIR + "/roary/number_of_genes_in_pan_genome.Rtab",
        OUTDIR + "/roary/number_of_new_genes.Rtab",
        OUTDIR + "/roary/number_of_unique_genes.Rtab",
        OUTDIR + "/roary/summary_statistics.txt"

rule preprocess_patric_gff:
    input:
        PATRICDATA + "/{species}/{id}/{id}.PATRIC.gff"
    output:
        OUTDIR + "/preprocess/{species}/{id}/{id}.gff"
    shell:
        # "perl -e 'while(<>){{chomp; "
        # "my ($chr, $pat, $type, $start, $end, $point, "
        # "$strand, $num, $annot) = split(/\t/, $_); "
        # "$chr =~ s/^accn\|//; my ($id, @trash) = split(/;/, $annot); "
        # "$id =~ s/^ID=fig\|//; print join(\"\t\", ($chr, $pat, $type, "
        # "$start, $end, $point, $strand, $num, $id)) . \"\n\"}}' "
        # "{input} > {output}"
        BINDIR + '/patric2roary_gff.py --infile {input} --outfile {output}'

rule preprocess_patric_fna:
    input:
        PATRICDATA + "/{species}/{id}/{id}.fna"
    output:
        OUTDIR + "/preprocess/{species}/{id}/{id}.fna"
    shell:
        BINDIR + '/fasta_remove_desc.py --infile {input} --outfile {output}'

rule create_patric_gff:
    input:
        OUTDIR + "/preprocess/{species}/{id}/{id}.gff",
        OUTDIR + "/preprocess/{species}/{id}/{id}.fna"
    output:
        OUTDIR + "/roary_input/{species}-{id}.gff"
    shell:
        "cat {input} > {output}"

rule roary:
    input:
        gffs = expand(OUTDIR + "/roary_input/{species}-{id}.gff", zip,
                      species=specs, id=ids)
    output:
        OUTDIR + "/roary/accessory_binary_genes.fa",
        OUTDIR + "/roary/accessory_graph.dot",
        OUTDIR + "/roary/accessory.header.embl",
        OUTDIR + "/roary/accessory.tab",
        OUTDIR + "/roary/blast_identity_frequency.Rtab",
        OUTDIR + "/roary/clustered_proteins",
        OUTDIR + "/roary/core_accessory_graph.dot",
        OUTDIR + "/roary/core_accessory.header.embl",
        OUTDIR + "/roary/core_accessory.tab",
        OUTDIR + "/roary/gene_presence_absence.csv",
        OUTDIR + "/roary/gene_presence_absence.Rtab",
        OUTDIR + "/roary/number_of_conserved_genes.Rtab",
        OUTDIR + "/roary/number_of_genes_in_pan_genome.Rtab",
        OUTDIR + "/roary/number_of_new_genes.Rtab",
        OUTDIR + "/roary/number_of_unique_genes.Rtab",
        OUTDIR + "/roary/summary_statistics.txt",
        dir = OUTDIR + "/roary/"
    shell:
        "roary -v -f {output.dir} {input.gffs}"
