{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from MKtest import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = dict(A=1, B=2, C=3, D=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Dn</th>\n",
       "      <th>Ds</th>\n",
       "      <th>Pn</th>\n",
       "      <th>Ps</th>\n",
       "      <th>hg</th>\n",
       "      <th>hg.pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1156433.3.peg.1</td>\n",
       "      <td>CP006776</td>\n",
       "      <td>1134</td>\n",
       "      <td>2262</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>29</td>\n",
       "      <td>71</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.037896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1156433.3.peg.2</td>\n",
       "      <td>CP006776</td>\n",
       "      <td>2271</td>\n",
       "      <td>2427</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>0.352381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1156433.3.peg.3</td>\n",
       "      <td>CP006776</td>\n",
       "      <td>2531</td>\n",
       "      <td>2897</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.447368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1156433.3.peg.4</td>\n",
       "      <td>CP006776</td>\n",
       "      <td>2935</td>\n",
       "      <td>4101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1156433.3.peg.5</td>\n",
       "      <td>CP006776</td>\n",
       "      <td>4162</td>\n",
       "      <td>4562</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>0.424242</td>\n",
       "      <td>0.622079</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene    contig  start   end  Dn  Ds  Pn  Ps        hg   hg.pval\n",
       "0  1156433.3.peg.1  CP006776   1134  2262   6   3  29  71  0.238095  0.037896\n",
       "1  1156433.3.peg.2  CP006776   2271  2427   0   0   2  15  0.187500  0.352381\n",
       "2  1156433.3.peg.3  CP006776   2531  2897   0   0   3  13  0.285714  0.447368\n",
       "3  1156433.3.peg.4  CP006776   2935  4101   1   1   3   1  2.000000  1.000000\n",
       "4  1156433.3.peg.5  CP006776   4162  4562   2   0  13  10  0.424242  0.622079"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = \"/home/sur/micropopgen/exp/today/mktest/Streptococcus_sp_57750/mk_results.Buccal.mucosa_Tongue.dorsum.txt\"\n",
    "pd.read_csv(file, sep=\"\\t\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ARGS:\n",
    "    def __init__(self, indir, metadata_file,\n",
    "                group1, group2, test='hg',\n",
    "                outfile='mk_results', min_count=1,\n",
    "                nrows=float('inf'), tables='mk_tables',\n",
    "                pseudocount=0, permutations=0, seed=5):\n",
    "        \n",
    "        self.indir = indir\n",
    "        self.metadata_file = metadata_file\n",
    "        self.group1 = group1\n",
    "        self.group2 = group2\n",
    "        self.test = test\n",
    "        self.outfile = outfile\n",
    "        self.min_count = min_count\n",
    "        self.nrows = nrows\n",
    "        self.tables = tables\n",
    "        self.pseudocount = pseudocount\n",
    "        self.permutations = permutations\n",
    "        self.seed = seed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = ARGS(indir='/home/sur/micropopgen/data/test_data/midas/merged.snps//Veillonella_parvula_57794/',\n",
    "            metadata_file='/home/sur/micropopgen/data/test_data/midas/map.txt',\n",
    "            group1='Supragingival plaque',\n",
    "            group2='Tongue dorsum',\n",
    "            outfile='mktest/Veillonella_parvula_57794//mk_results.Supragingival.plaque_Tongue.dorsum.txt',\n",
    "            tables='mktest/Veillonella_parvula_57794//mk_tables.Supragingival.plaque_Tongue.dorsum.txt',\n",
    "            nrows=2000,\n",
    "            pseudocount=0,\n",
    "            permutations=10,\n",
    "            seed=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking MIDAS files exist\n",
      "\tAll files found\n",
      "Read metadata\n",
      "Calculate MK contingency tables\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1830\n",
      "Number of genes: 3\n",
      "Sites with counts: 1830\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1830\n",
      "Number of genes: 3\n",
      "Sites with counts: 1830\n",
      "Genes with MK: 3\n"
     ]
    }
   ],
   "source": [
    "print(\"Checking MIDAS files exist\")\n",
    "confirm_midas_merge_files(args)\n",
    "\n",
    "# Read mapping files\n",
    "# Create dictionaries that have all the samples per group (Groups),\n",
    "# and the group to which each sample belongs (Samples)\n",
    "# Probably should change this to pandas\n",
    "print(\"Read metadata\")\n",
    "Samples, Groups = process_metadata_file(args.metadata_file)\n",
    "\n",
    "print(\"Calculate MK contingency tables\")\n",
    "MK, Genes = calculate_contingency_tables(Samples, Groups, args)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Permuting\n",
      "Seed is 5\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1876\n",
      "Number of genes: 3\n",
      "Sites with counts: 1876\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1876\n",
      "Number of genes: 3\n",
      "Sites with counts: 1876\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n",
      "\tRead snps_info.txt\n",
      "============HEADERs============\n",
      ">Site id: site_id\n",
      ">Contig: ref_id\n",
      ">Position: ref_pos\n",
      ">Ref allele: ref_allele\n",
      ">Major allele: major_allele\n",
      ">Minor allele: minor_allele\n",
      ">Locus type: locus_type\n",
      ">Gene id: gene_id\n",
      ">Aminoacids: amino_acids\n",
      "\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "\tChose sites based on depth in groups to compare\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "\tRead frequencies and calculate\n",
      "Processing snp_freq.txt\n",
      "Number of sites: 1883\n",
      "Number of genes: 3\n",
      "Sites with counts: 1883\n",
      "Genes with MK: 3\n"
     ]
    }
   ],
   "source": [
    "if args.permutations > 0:\n",
    "        MK = [MK]\n",
    "        print(\"Permuting\")\n",
    "        print(\"Seed is {}\".format(str(args.seed)))\n",
    "        for i in range(args.permutations):\n",
    "            Sp, Gp = process_metadata_file(args.metadata_file, permute=True)\n",
    "            MK.append(calculate_contingency_tables(Sp, Gp, args))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['hg']\n",
      "['hg.pval']\n"
     ]
    }
   ],
   "source": [
    "# Get list of tests to perform\n",
    "test = args.test\n",
    "supported_tests = ['NI', 'ratio', 'hg',\n",
    "                   'G', 'G_Yates', 'G_Williams_p',\n",
    "                   'alpha']\n",
    "if test == 'all':\n",
    "    test = supported_tests\n",
    "elif test not in supported_tests:\n",
    "    raise ValueError(\"Test not supported\")\n",
    "else:\n",
    "    test = [test]\n",
    "\n",
    "# Create header\n",
    "header_base = ['gene', 'contig', 'start', 'end',\n",
    "               'Dn', 'Ds', 'Pn', 'Ps']\n",
    "pval_list = [''.join([t, '.pval']) for t in test]\n",
    "header = header_base + test + pval_list\n",
    "print(test)\n",
    "print(pval_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'686660.3.peg.1': <MKtest.MKtest at 0x7f96002e3c50>,\n",
       " '686660.3.peg.2': <MKtest.MKtest at 0x7f96002e3a20>,\n",
       " '686660.3.peg.3': <MKtest.MKtest at 0x7f96002e3748>}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
