{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# The purpose of these notebook is to create \n",
    "# The skeleton of a database, and populate it with HMP data\n",
    "import sqlite3\n",
    "# import csv\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Filenames\n",
    "sample_file = '/home/sur/micropopgen/data/HMP/hmp_catalogue_wgs_all_metadata.csv'\n",
    "runs_file = '/home/sur/micropopgen/data/HMP/2017-07-17.all_runs/all_runs.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Read Archive ID</th>\n",
       "      <th>NCBI Taxon ID</th>\n",
       "      <th>NCBI Project ID</th>\n",
       "      <th>HMP Isolation Body Site</th>\n",
       "      <th>HMP Isolation Body Subsite</th>\n",
       "      <th>HMP Body Product</th>\n",
       "      <th>Isolation Source</th>\n",
       "      <th>Project Status</th>\n",
       "      <th>Gene Count</th>\n",
       "      <th>Reads Count</th>\n",
       "      <th>...</th>\n",
       "      <th>Funding Source</th>\n",
       "      <th>Sequencing Method</th>\n",
       "      <th>Assembly Methods</th>\n",
       "      <th>Gene Calling Methods</th>\n",
       "      <th>Addition Date</th>\n",
       "      <th>Last Modification Date</th>\n",
       "      <th>Host Gender</th>\n",
       "      <th>MRN (Subject ID)</th>\n",
       "      <th>Visit Number</th>\n",
       "      <th>Replicate Number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRS011061</td>\n",
       "      <td>646099</td>\n",
       "      <td>48479</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>173933</td>\n",
       "      <td>77700000</td>\n",
       "      <td>...</td>\n",
       "      <td>NIH-HMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011-04-25T20:00:00</td>\n",
       "      <td>2011-08-15T20:00:00</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRS011084</td>\n",
       "      <td>646099</td>\n",
       "      <td>48479</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>0</td>\n",
       "      <td>141800000</td>\n",
       "      <td>...</td>\n",
       "      <td>NIH-HMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011-04-25T20:00:00</td>\n",
       "      <td>2011-08-15T20:00:00</td>\n",
       "      <td>Male</td>\n",
       "      <td>158479027</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRS011086</td>\n",
       "      <td>646099</td>\n",
       "      <td>48479</td>\n",
       "      <td>oral</td>\n",
       "      <td>Tongue dorsum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>86427</td>\n",
       "      <td>65200000</td>\n",
       "      <td>...</td>\n",
       "      <td>NIH-HMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011-04-25T20:00:00</td>\n",
       "      <td>2011-08-15T20:00:00</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRS011090</td>\n",
       "      <td>646099</td>\n",
       "      <td>48479</td>\n",
       "      <td>oral</td>\n",
       "      <td>Buccal mucosa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>9445</td>\n",
       "      <td>74100000</td>\n",
       "      <td>...</td>\n",
       "      <td>NIH-HMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011-04-25T20:00:00</td>\n",
       "      <td>2011-08-15T20:00:00</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRS011098</td>\n",
       "      <td>646099</td>\n",
       "      <td>48479</td>\n",
       "      <td>oral</td>\n",
       "      <td>Supragingival plaque</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>120288</td>\n",
       "      <td>137800000</td>\n",
       "      <td>...</td>\n",
       "      <td>NIH-HMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011-04-25T20:00:00</td>\n",
       "      <td>2011-08-15T20:00:00</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence Read Archive ID  NCBI Taxon ID  NCBI Project ID  \\\n",
       "0                SRS011061         646099            48479   \n",
       "1                SRS011084         646099            48479   \n",
       "2                SRS011086         646099            48479   \n",
       "3                SRS011090         646099            48479   \n",
       "4                SRS011098         646099            48479   \n",
       "\n",
       "  HMP Isolation Body Site HMP Isolation Body Subsite HMP Body Product  \\\n",
       "0  gastrointestinal_tract                        NaN            Stool   \n",
       "1  gastrointestinal_tract                        NaN            Stool   \n",
       "2                    oral              Tongue dorsum              NaN   \n",
       "3                    oral              Buccal mucosa              NaN   \n",
       "4                    oral       Supragingival plaque              NaN   \n",
       "\n",
       "  Isolation Source Project Status  Gene Count  Reads Count       ...         \\\n",
       "0              NaN       Complete      173933     77700000       ...          \n",
       "1              NaN       Complete           0    141800000       ...          \n",
       "2              NaN       Complete       86427     65200000       ...          \n",
       "3              NaN       Complete        9445     74100000       ...          \n",
       "4              NaN       Complete      120288    137800000       ...          \n",
       "\n",
       "   Funding Source  Sequencing Method  Assembly Methods Gene Calling Methods  \\\n",
       "0         NIH-HMP                NaN               NaN                  NaN   \n",
       "1         NIH-HMP                NaN               NaN                  NaN   \n",
       "2         NIH-HMP                NaN               NaN                  NaN   \n",
       "3         NIH-HMP                NaN               NaN                  NaN   \n",
       "4         NIH-HMP                NaN               NaN                  NaN   \n",
       "\n",
       "         Addition Date Last Modification Date Host Gender  MRN (Subject ID)  \\\n",
       "0  2011-04-25T20:00:00    2011-08-15T20:00:00      Female         158458797   \n",
       "1  2011-04-25T20:00:00    2011-08-15T20:00:00        Male         158479027   \n",
       "2  2011-04-25T20:00:00    2011-08-15T20:00:00      Female         158458797   \n",
       "3  2011-04-25T20:00:00    2011-08-15T20:00:00      Female         158458797   \n",
       "4  2011-04-25T20:00:00    2011-08-15T20:00:00      Female         158458797   \n",
       "\n",
       "   Visit Number Replicate Number  \n",
       "0             1                0  \n",
       "1             1                0  \n",
       "2             1                0  \n",
       "3             1                0  \n",
       "4             1                0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read sample metadata\n",
    "samples = pd.read_csv(sample_file)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA48479</td>\n",
       "      <td>SRP002163</td>\n",
       "      <td>SAMN00032498</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>SRX023565</td>\n",
       "      <td>SRR060370</td>\n",
       "      <td>SRA020838</td>\n",
       "      <td>646099</td>\n",
       "      <td>human metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700014562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human metagenome sample from G_DNA_Stool of a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA48479</td>\n",
       "      <td>SRP002163</td>\n",
       "      <td>SAMN00032498</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>SRX023566</td>\n",
       "      <td>SRR060371</td>\n",
       "      <td>SRA020838</td>\n",
       "      <td>646099</td>\n",
       "      <td>human metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700014562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human metagenome sample from G_DNA_Stool of a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA48479</td>\n",
       "      <td>SRP002163</td>\n",
       "      <td>SAMN00032521</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRX024515</td>\n",
       "      <td>SRR061903</td>\n",
       "      <td>SRA020838</td>\n",
       "      <td>646099</td>\n",
       "      <td>human metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700014724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human metagenome sample from G_DNA_Stool of a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA48479</td>\n",
       "      <td>SRP002163</td>\n",
       "      <td>SAMN00032521</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRX024516</td>\n",
       "      <td>SRR061904</td>\n",
       "      <td>SRA020838</td>\n",
       "      <td>646099</td>\n",
       "      <td>human metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700014724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human metagenome sample from G_DNA_Stool of a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA48479</td>\n",
       "      <td>SRP002163</td>\n",
       "      <td>SAMN00032521</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRX024714</td>\n",
       "      <td>SRR062102</td>\n",
       "      <td>SRA020838</td>\n",
       "      <td>646099</td>\n",
       "      <td>human metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700014724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human metagenome sample from G_DNA_Stool of a ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession secondary_study_accession sample_accession  \\\n",
       "0      PRJNA48479                 SRP002163     SAMN00032498   \n",
       "1      PRJNA48479                 SRP002163     SAMN00032498   \n",
       "2      PRJNA48479                 SRP002163     SAMN00032521   \n",
       "3      PRJNA48479                 SRP002163     SAMN00032521   \n",
       "4      PRJNA48479                 SRP002163     SAMN00032521   \n",
       "\n",
       "  secondary_sample_accession experiment_accession run_accession  \\\n",
       "0                  SRS011061            SRX023565     SRR060370   \n",
       "1                  SRS011061            SRX023566     SRR060371   \n",
       "2                  SRS011084            SRX024515     SRR061903   \n",
       "3                  SRS011084            SRX024516     SRR061904   \n",
       "4                  SRS011084            SRX024714     SRR062102   \n",
       "\n",
       "  submission_accession  tax_id   scientific_name instrument_platform  \\\n",
       "0            SRA020838  646099  human metagenome            ILLUMINA   \n",
       "1            SRA020838  646099  human metagenome            ILLUMINA   \n",
       "2            SRA020838  646099  human metagenome            ILLUMINA   \n",
       "3            SRA020838  646099  human metagenome            ILLUMINA   \n",
       "4            SRA020838  646099  human metagenome            ILLUMINA   \n",
       "\n",
       "     ...     sra_ftp sra_aspera sra_galaxy  cram_index_ftp cram_index_aspera  \\\n",
       "0    ...         NaN        NaN        NaN             NaN               NaN   \n",
       "1    ...         NaN        NaN        NaN             NaN               NaN   \n",
       "2    ...         NaN        NaN        NaN             NaN               NaN   \n",
       "3    ...         NaN        NaN        NaN             NaN               NaN   \n",
       "4    ...         NaN        NaN        NaN             NaN               NaN   \n",
       "\n",
       "  cram_index_galaxy sample_alias  broker_name  \\\n",
       "0               NaN    700014562          NaN   \n",
       "1               NaN    700014562          NaN   \n",
       "2               NaN    700014724          NaN   \n",
       "3               NaN    700014724          NaN   \n",
       "4               NaN    700014724          NaN   \n",
       "\n",
       "                                        sample_title subject_id  \n",
       "0  Human metagenome sample from G_DNA_Stool of a ...        NaN  \n",
       "1  Human metagenome sample from G_DNA_Stool of a ...        NaN  \n",
       "2  Human metagenome sample from G_DNA_Stool of a ...        NaN  \n",
       "3  Human metagenome sample from G_DNA_Stool of a ...        NaN  \n",
       "4  Human metagenome sample from G_DNA_Stool of a ...        NaN  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read run metadata\n",
    "runs = pd.read_csv(runs_file,sep='\\t')\n",
    "runs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    SRS011061\n",
      "1    SRS011084\n",
      "2    SRS011086\n",
      "3    SRS011090\n",
      "4    SRS011098\n",
      "Name: Sequence Read Archive ID, dtype: object\n",
      "0    SRS011061\n",
      "1    SRS011061\n",
      "2    SRS011084\n",
      "3    SRS011084\n",
      "4    SRS011084\n",
      "Name: secondary_sample_accession, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Checking columns that have SRA sample ID (ie SRS###)\n",
    "print(samples['Sequence Read Archive ID'].head())\n",
    "print(runs['secondary_sample_accession'].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Dropping unnecessary columns from pandas data frames\n",
    "samples_small = samples.drop(['NCBI Taxon ID', 'NCBI Project ID',\n",
    "                              'Project Status', 'Gene Count',\n",
    "                              'Reads Count', 'Size (KB)',\n",
    "                              'GC Content', 'IMG/HMP ID',\n",
    "                              'Sequencing Center',\n",
    "                              'Sequencing Country',\n",
    "                              'Funding Source', 'Sequencing Method',\n",
    "                              'Assembly Methods','Gene Calling Methods',\n",
    "                              'Addition Date',\n",
    "                              'Last Modification Date',\n",
    "                              'Isolation Source',\n",
    "                              'HMP Body Product'],\n",
    "                             axis = 1)\n",
    "runs_small = runs.drop(['study_accession',\n",
    "                        'secondary_study_accession',\n",
    "                        'sample_accession',\n",
    "                        'experiment_accession',\n",
    "                        'study_title','study_alias',\n",
    "                        'submission_accession','tax_id',\n",
    "                        'scientific_name','instrument_platform',\n",
    "                        'library_source','library_selection',\n",
    "                        'read_count','base_count','center_name',\n",
    "                        'first_public','last_updated',\n",
    "                        'nominal_length','library_strategy',\n",
    "                        'run_alias','fastq_bytes','fastq_md5',\n",
    "                        'fastq_ftp','fastq_aspera',\n",
    "                        'fastq_galaxy','submitted_bytes',\n",
    "                        'submitted_md5','experiment_alias',\n",
    "                        'submitted_ftp','submitted_aspera',\n",
    "                        'submitted_galaxy','submitted_format',\n",
    "                        'sra_bytes','sra_md5','sra_ftp',\n",
    "                        'sra_aspera','sra_galaxy',\n",
    "                        'cram_index_ftp','cram_index_aspera',\n",
    "                        'cram_index_galaxy','sample_alias',\n",
    "                        'broker_name','sample_title'],\n",
    "                       axis = 1)\n",
    "# print(samples_small.columns)\n",
    "# print(samples_small.head())\n",
    "# print(runs_small.columns)\n",
    "# print(runs_small.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 'Illumina Genome Analyzer II paired end sequencing; Deep WGS Genome Sequencing of HMP clinical samples',\n",
       "       'Illumina Genome Analyzer II sequencing; Deep WGS Genome Sequencing of HMP clinical samples',\n",
       "       'Illumina Genome Analyzer II paired end sequencing',\n",
       "       '454 GS FLX Titanium sequencing',\n",
       "       'Illumina HiSeq 2000 paired end sequencing',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; Illumina sequencing of Clinical Metagenomic Shotgun Production 764305738-700024545-zg_g_dna_stool paired end RANDOM library',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700098677-PE-1 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700098681-PE-1 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700107197-PE-01 Experiment ',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700102307-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700107195-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700107201-PE-01 Experiment',\n",
       "       'Illumina HiSeq 1000 paired end sequencing',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700107189-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700102305-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700102311-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700111032-PE-01 Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700111038-PE-01 Experiment Run',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700111026-PE-1  Experiment',\n",
       "       'Illumina Genome Analyzer II paired end sequencing; HUMAN METAGENOME 700111034-PE-1 Experiment',\n",
       "       'Illumina HiSeq 1000 paired end sequencing; Illumina Sequencing of Clinical Metagenomic sample for Human Microbiome Project'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that none of the titles wher subject id is missing has\n",
    "# any information\n",
    "# HAVE NOT CHECKED THAT SUBJECT IDs IN TITLE MATCH COLUMN\n",
    "runs_small.loc[ runs_small['subject_id'].isnull(), 'experiment_title'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gastrointestinal_tract    157\n",
       "oral                        5\n",
       "Name: HMP Isolation Body Site, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify the subsites without body site\n",
    "subsites = samples_small['HMP Isolation Body Subsite']\n",
    "subsite_missing = subsites.isnull()\n",
    "sites = samples_small.loc[subsite_missing,'HMP Isolation Body Site']\n",
    "sites.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NaN      157\n",
       "Stool     97\n",
       "Name: HMP Isolation Body Subsite, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check body subsite for all GI samples\n",
    "sites = samples_small['HMP Isolation Body Site']\n",
    "sites.value_counts(dropna=False)\n",
    "sites_gi = sites == 'gastrointestinal_tract'\n",
    "subsite = samples_small.loc[ sites_gi, 'HMP Isolation Body Subsite']\n",
    "subsite.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Stool    254\n",
       "Name: HMP Isolation Body Subsite, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All samples from GI tract are from stool, so I will change the NaN to that\n",
    "samples_small.loc[ samples_small['HMP Isolation Body Site'] == 'gastrointestinal_tract',\n",
    "                  'HMP Isolation Body Subsite'] = 'Stool'\n",
    "# Check body subsite for all GI samples\n",
    "sites = samples_small['HMP Isolation Body Site']\n",
    "sites.value_counts(dropna=False)\n",
    "sites_gi = sites == 'gastrointestinal_tract'\n",
    "subsite = samples_small.loc[ sites_gi, 'HMP Isolation Body Subsite']\n",
    "subsite.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# There are 5 oral samples that have no subsite. I will set them to 'oral'\n",
    "samples_small.loc[ samples_small['HMP Isolation Body Subsite'].isnull() &\n",
    "                  (samples_small['HMP Isolation Body Site'] == 'oral'),\n",
    "                  'HMP Isolation Body Subsite'] = 'oral'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "oral                      690\n",
       "gastrointestinal_tract    254\n",
       "airways                   138\n",
       "urogenital_tract          128\n",
       "skin                       55\n",
       "Name: HMP Isolation Body Site, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_small['HMP Isolation Body Site'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Stool                           254\n",
       "Tongue dorsum                   221\n",
       "Supragingival plaque            210\n",
       "Buccal mucosa                   185\n",
       "Nares                           138\n",
       "Posterior fornix                108\n",
       "Retroauricular crease            54\n",
       "Palatine tonsils                 20\n",
       "Subgingival plaque               19\n",
       "Throat                           13\n",
       "Vaginal                          12\n",
       "Vaginal introitus                 8\n",
       "Attached Keratinized gingiva      8\n",
       "Attached/Keratinized gingiva      6\n",
       "oral                              5\n",
       "Saliva                            2\n",
       "Hard palate                       1\n",
       "Antecubital fossa                 1\n",
       "Name: HMP Isolation Body Subsite, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_small['HMP Isolation Body Subsite'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "IntegrityError",
     "evalue": "UNIQUE constraint failed: sample.SRS",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIntegrityError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-efc0aff07c5d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m              r['Replicate Number'])\n\u001b[1;32m     13\u001b[0m     \u001b[0;31m#print(values)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m     \u001b[0mc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"insert into sample values (null,?,?,?,?,?,?,?)\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Sequence Read Archive ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIntegrityError\u001b[0m: UNIQUE constraint failed: sample.SRS"
     ]
    }
   ],
   "source": [
    "# Now I have homogenized the body subsites. I will fill the table\n",
    "conn = sqlite3.connect('/home/sur/micropopgen/data/database/metagenomes.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# Iterate and add\n",
    "for i,r in samples_small.head().iterrows():\n",
    "    #print(r)\n",
    "    values = (r['MRN (Subject ID)'],r['HMP Isolation Body Subsite'],\n",
    "             r['Visit Number'],'NULL',\n",
    "             r['Sequence Read Archive ID'],\n",
    "             'Homo sapiens',\n",
    "             r['Replicate Number'])\n",
    "    #print(values)\n",
    "    c.execute(\"insert into sample values (null,?,?,?,?,?,?,?)\", values)\n",
    "    print(r['Sequence Read Archive ID'])\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence Read Archive ID</th>\n",
       "      <th>HMP Isolation Body Site</th>\n",
       "      <th>HMP Isolation Body Subsite</th>\n",
       "      <th>Host Gender</th>\n",
       "      <th>MRN (Subject ID)</th>\n",
       "      <th>Visit Number</th>\n",
       "      <th>Replicate Number</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRS011061</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>Stool</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>SRR060370</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>Solexa-27630</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina Genome Analyzer II paired end sequenc...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRS011061</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>Stool</td>\n",
       "      <td>Female</td>\n",
       "      <td>158458797</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>SRR060371</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>Solexa-27630</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina Genome Analyzer II paired end sequenc...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRS011084</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>Stool</td>\n",
       "      <td>Male</td>\n",
       "      <td>158479027</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRR061903</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>Solexa-28680</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina Genome Analyzer II paired end sequenc...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRS011084</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>Stool</td>\n",
       "      <td>Male</td>\n",
       "      <td>158479027</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRR061904</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>Solexa-28680</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina Genome Analyzer II paired end sequenc...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRS011084</td>\n",
       "      <td>gastrointestinal_tract</td>\n",
       "      <td>Stool</td>\n",
       "      <td>Male</td>\n",
       "      <td>158479027</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>SRR062102</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>Solexa-28680</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina Genome Analyzer II paired end sequenc...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence Read Archive ID HMP Isolation Body Site HMP Isolation Body Subsite  \\\n",
       "0                SRS011061  gastrointestinal_tract                      Stool   \n",
       "1                SRS011061  gastrointestinal_tract                      Stool   \n",
       "2                SRS011084  gastrointestinal_tract                      Stool   \n",
       "3                SRS011084  gastrointestinal_tract                      Stool   \n",
       "4                SRS011084  gastrointestinal_tract                      Stool   \n",
       "\n",
       "  Host Gender  MRN (Subject ID)  Visit Number  Replicate Number  \\\n",
       "0      Female         158458797             1                 0   \n",
       "1      Female         158458797             1                 0   \n",
       "2        Male         158479027             1                 0   \n",
       "3        Male         158479027             1                 0   \n",
       "4        Male         158479027             1                 0   \n",
       "\n",
       "  secondary_sample_accession run_accession             instrument_model  \\\n",
       "0                  SRS011061     SRR060370  Illumina Genome Analyzer II   \n",
       "1                  SRS011061     SRR060371  Illumina Genome Analyzer II   \n",
       "2                  SRS011084     SRR061903  Illumina Genome Analyzer II   \n",
       "3                  SRS011084     SRR061904  Illumina Genome Analyzer II   \n",
       "4                  SRS011084     SRR062102  Illumina Genome Analyzer II   \n",
       "\n",
       "   library_name library_layout  \\\n",
       "0  Solexa-27630         PAIRED   \n",
       "1  Solexa-27630         PAIRED   \n",
       "2  Solexa-28680         PAIRED   \n",
       "3  Solexa-28680         PAIRED   \n",
       "4  Solexa-28680         PAIRED   \n",
       "\n",
       "                                    experiment_title  subject_id  \n",
       "0  Illumina Genome Analyzer II paired end sequenc...         NaN  \n",
       "1  Illumina Genome Analyzer II paired end sequenc...         NaN  \n",
       "2  Illumina Genome Analyzer II paired end sequenc...         NaN  \n",
       "3  Illumina Genome Analyzer II paired end sequenc...         NaN  \n",
       "4  Illumina Genome Analyzer II paired end sequenc...         NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#############\n",
    "# Now we move to the runs table\n",
    "# Merge tables to get subject ID\n",
    "merged = pd.merge(samples_small, runs_small,\n",
    "                  right_on='secondary_sample_accession',\n",
    "                  left_on = 'Sequence Read Archive ID')\n",
    "merged.head()\n",
    "# runs_small.head()\n",
    "# merged.columns\n",
    "# print(merged['MRN (Subject ID)'].head())\n",
    "# print(merged['subject_id'].head())\n",
    "# print(merged.loc[ merged['subject_id'].notnull(), 'subject_id'].head())\n",
    "# print(merged.loc[ merged['subject_id'].notnull(), 'MRN (Subject ID)'].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR060370\n",
      "SRR060371\n",
      "SRR061903\n",
      "SRR061904\n",
      "SRR062102\n"
     ]
    }
   ],
   "source": [
    "conn = sqlite3.connect('/home/sur/micropopgen/data/database/metagenomes.db')\n",
    "c = conn.cursor()\n",
    "# Iterate and add\n",
    "for i,r in merged.head().iterrows():\n",
    "    #print(r)\n",
    "    values = (r['secondary_sample_accession'],\n",
    "              r['MRN (Subject ID)'],\n",
    "              r['instrument_model'],\n",
    "              r['library_layout'],\n",
    "              r['run_accession'])\n",
    "    #print(values)\n",
    "    c.execute(\"insert into run values (null,?,?,?,?,?)\", values)\n",
    "    print(r['run_accession'])\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>body_site</th>\n",
       "      <th>visit</th>\n",
       "      <th>age</th>\n",
       "      <th>SRS</th>\n",
       "      <th>host_species</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>158458797</td>\n",
       "      <td>Stool</td>\n",
       "      <td>1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>SRS011061</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>158479027</td>\n",
       "      <td>Stool</td>\n",
       "      <td>1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>SRS011084</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>158458797</td>\n",
       "      <td>Tongue dorsum</td>\n",
       "      <td>1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>SRS011086</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>158458797</td>\n",
       "      <td>Buccal mucosa</td>\n",
       "      <td>1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>SRS011090</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>158458797</td>\n",
       "      <td>Supragingival plaque</td>\n",
       "      <td>1</td>\n",
       "      <td>NULL</td>\n",
       "      <td>SRS011098</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id  subject_id             body_site  visit   age        SRS  \\\n",
       "0          1   158458797                 Stool      1  NULL  SRS011061   \n",
       "1          2   158479027                 Stool      1  NULL  SRS011084   \n",
       "2          3   158458797         Tongue dorsum      1  NULL  SRS011086   \n",
       "3          4   158458797         Buccal mucosa      1  NULL  SRS011090   \n",
       "4          5   158458797  Supragingival plaque      1  NULL  SRS011098   \n",
       "\n",
       "   host_species  replicate  \n",
       "0  Homo sapiens          0  \n",
       "1  Homo sapiens          0  \n",
       "2  Homo sapiens          0  \n",
       "3  Homo sapiens          0  \n",
       "4  Homo sapiens          0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn = sqlite3.connect('/home/sur/micropopgen/data/database/metagenomes.db')\n",
    "\n",
    "#samples_small.head()\n",
    "pd.read_sql_query(\"select * from sample limit 10;\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#samples_small.head()\n",
    "pd.read_sql_query(\"select * from run limit 20;\", conn)\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
