{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# Copyright (C) 2017 Sur Herrera Paredes\n",
    "\n",
    "# Imports\n",
    "import os\n",
    "import sutilspy\n",
    "import csv\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Arguments for ipython\n",
    "args = argparse.Namespace()\n",
    "args.indir = \"/home/sur/micropopgen/exp/2017/today5/test/\"\n",
    "args.test = \"MK\"\n",
    "args.outfile = \"/home/sur/micropopgen/exp/2017/today5/mk_results.txt\"\n",
    "args.metadata_file = \"/home/sur/micropopgen/exp/2017/today5/map.txt\"\n",
    "args.group1 = \"Supragingival plaque\"\n",
    "args.group2 = \"Tongue dorsum\"\n",
    "args.min_count = 3\n",
    "args.nrow = 30\n",
    "\n",
    "# indir = 'snps/'\n",
    "# test = 'MK'\n",
    "# outfile = 'mk_results.txt'\n",
    "# metadata_file = 'map.txt'\n",
    "# group1 = 'A'\n",
    "# group2 = 'B'\n",
    "# min_count = 3\n",
    "# nrows = 30000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class GenomeSite:\n",
    "    \"\"\"A class for represintinc sites in genome that have potential SNPS\"\"\"\n",
    "    \n",
    "    def __init__(self,site_id, contig, position, ref_allele = '',\n",
    "                 major_allele = '',\n",
    "                 minor_allele = '', locus_type = '', gene_id = '',\n",
    "                 aminoacid_A = '',\n",
    "                 aminoacid_C = '', aminoacid_G = '', aminoacid_T = ''):\n",
    "        self.id = site_id\n",
    "        self.contig = contig\n",
    "        self.position = position\n",
    "        self.ref_allele = ref_allele\n",
    "        self.major_allele = major_allele\n",
    "        self.minor_allele = minor_allele\n",
    "        self.locus_type = locus_type\n",
    "        self.gene_id = gene_id\n",
    "        self.aminoA = aminoacid_A\n",
    "        self.aminoC = aminoacid_C\n",
    "        self.aminoG = aminoacid_G\n",
    "        self.aminoT = aminoacid_T\n",
    "    \n",
    "    def codon_aminoacid(self, base):\n",
    "        if base in ['A','a']:\n",
    "            return(self.aminoA)\n",
    "        elif base in ['C','c']:\n",
    "            return(self.aminoA)\n",
    "        elif base in ['G','g']:\n",
    "            return(self.aminoA)\n",
    "        elif base in ['T','t']:\n",
    "            return(self.aminoA)\n",
    "        else:\n",
    "            raise ValueError(\"base must be one of the four canonical nucleoties\")\n",
    "    \n",
    "    def substitution_type(self):\n",
    "        substitution_type = ''\n",
    "        if self.codon_aminoacid(base = self.major_allele) == self.codon_aminoacid(base = self.minor_allele):\n",
    "            substitution_type = 'synonymous'\n",
    "        else:\n",
    "            substitution_type = 'non-synonymous'\n",
    "        \n",
    "        return(substitution_type)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Gene:\n",
    "    \"\"\"A class for representing a gene\"\"\"\n",
    "    \n",
    "    def __init__(self, gene_id,contig,start,end, strand = ''):\n",
    "        if(start > end):\n",
    "            raise ValueError(\"Start cannot be greater than end\")\n",
    "        self.id = gene_id\n",
    "        self.contig = contig\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "    \n",
    "    def extend(self, pos):\n",
    "        if pos > self.end:\n",
    "            self.end = pos\n",
    "        elif pos < self.start:\n",
    "            self.start = pos\n",
    "    \n",
    "    def print(self):\n",
    "        print(\"===Gene===\")\n",
    "        print(\">Gene id: {}\".format(self.id))\n",
    "        print(\">Gene start: {}\".format(str(self.start)))\n",
    "        print(\">Gene end: {}\".format(str(self.end)))\n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class MKtest:\n",
    "    \"\"\"A class for holding the McDonald-Kreitmant test\"\"\"\n",
    "    \n",
    "    def __init__(self, name, Ds = 0, Dn = 0, Ps = 0, Pn = 0):\n",
    "        self.name = name\n",
    "        self.Dn = Dn\n",
    "        self.Ds = Ds\n",
    "        self.Ps = Ps\n",
    "        self.Pn = Pn\n",
    "    \n",
    "    def update(self, Ds = 0, Dn = 0, Ps = 0, Pn = 0):\n",
    "        \"\"\"Update the contigency matrix\"\"\"\n",
    "        self.Dn += Dn\n",
    "        self.Ds += Ds\n",
    "        self.Ps += Ps\n",
    "        self.Pn += Pn\n",
    "    \n",
    "    def calculate(self):\n",
    "        \"\"\"Calculate the McDonald Kreitman ratio\"\"\"\n",
    "        ratio = (self.Dn / self.Ds) / (self.Pn / self.Ps)\n",
    "        return(ratio)\n",
    "    \n",
    "    def alpha(self):\n",
    "        \"\"\"Calculate the Smith & Eyre-Walker alpha\"\"\"\n",
    "        alpha = 1 - ((self.Ds * self.Dn) / (self.Ps * self.Pn))\n",
    "        return(alpha)\n",
    "    \n",
    "    def test(self):\n",
    "        res = stats.fisher_exact([[self.Ds,self.Ps],[self.Dn,self.Pn]])\n",
    "        return(res)\n",
    "    \n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check files exist in input directory\n",
    "# Convert into function\n",
    "file_list = os.listdir(args.indir)\n",
    "if 'snps_freq.txt' not in file_list:\n",
    "    raise FileNotFoundError(\"Could not find snps_freq.txt at {}\".format(args.indir))\n",
    "if 'snps_info.txt' not in file_list:\n",
    "    raise FileNotFoundError(\"Could not find snps_info.txt at {}\".format(args.indir))\n",
    "if 'snps_depth.txt' not in file_list:\n",
    "    raise FileNotFoundError(\"Could not find snps_depth.txt at {}\".format(args.indir))\n",
    "if not os.path.isfile(args.metadata_file):\n",
    "    raise FileNotFoundError(\"Could not find metadata file {}\".format(args.metadata_file))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=============================================\n",
      "> Processing map of runs\n",
      "\tProcessed 54 runs in 3 samples\n",
      "=============================================\n",
      "\n",
      "=============================================\n",
      "> Processing map of runs\n",
      "\tProcessed 54 runs in 54 samples\n",
      "=============================================\n"
     ]
    }
   ],
   "source": [
    "# Read metadata\n",
    "Groups = sutilspy.io.process_run_list(args.metadata_file, 1, 0, header = True)\n",
    "Samples = sutilspy.io.process_run_list(args.metadata_file, 0, 1, header = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'indir' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-387780124f41>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mGenes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mSites\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindir\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'/snps_info.txt'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0minfo_fh\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minfo_fh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'indir' is not defined"
     ]
    }
   ],
   "source": [
    "# Read info\n",
    "Genes = {}\n",
    "Sites = {}\n",
    "with open(indir + '/snps_info.txt') as info_fh:\n",
    "    header = info_fh.readline()\n",
    "    header = header.split('\\t')\n",
    "    print(header)\n",
    "    info_reader = csv.reader(info_fh, delimiter = '\\t')\n",
    "    i = 0\n",
    "    \n",
    "    # Set columns\n",
    "    site_id_col = 0\n",
    "    contig_col = 1\n",
    "    pos_col = 2\n",
    "    ref_allele_col = 3\n",
    "    major_allele_col = 4\n",
    "    minor_allele_col = 5\n",
    "    locus_type_col = 11\n",
    "    gene_id_col = 12\n",
    "    aminoacids_col = 15\n",
    "    \n",
    "    print(\"============HEADERs============\")\n",
    "    print(\">Site id: {}\".format(header[site_id_col]))\n",
    "    print(\">Contig: {}\".format(header[contig_col]))\n",
    "    print(\">Position: {}\".format(header[pos_col]))\n",
    "    print(\">Ref allele: {}\".format(header[ref_allele_col]))\n",
    "    print(\">Major allele: {}\".format(header[major_allele_col]))\n",
    "    print(\">Minor allele: {}\".format(header[minor_allele_col]))\n",
    "    print(\">Locus type: {}\".format(header[locus_type_col]))\n",
    "    print(\">Gene id: {}\".format(header[gene_id_col]))\n",
    "    print(\">Aminoacids: {}\".format(header[aminoacids_col]))\n",
    "    \n",
    "    #\n",
    "    for row in info_reader:\n",
    "        i += 1\n",
    "        #print(row)\n",
    "        #print(row[gene_id_col], row[site_id_col])\n",
    "        #print(row[aminoacids_col])\n",
    "        gene = row[gene_id_col]\n",
    "        site_id = row[site_id_col]\n",
    "        aminoacids = row[aminoacids_col]\n",
    "        #print(aminoacids)\n",
    "        \n",
    "        if gene == 'NA':\n",
    "            # skip intergenig regions\n",
    "            continue\n",
    "        \n",
    "        # Get aminoacid per position\n",
    "        aa = aminoacids.split(',')\n",
    "        #print(aa)\n",
    "        \n",
    "        # Define site\n",
    "        Sites[site_id] = GenomeSite(site_id = site_id,\n",
    "                                    contig = row[contig_col],\n",
    "                                    position = row[pos_col],\n",
    "                                    ref_allele = row[ref_allele_col],\n",
    "                                    major_allele = row[major_allele_col],\n",
    "                                    minor_allele = row[minor_allele_col],\n",
    "                                    locus_type = row[locus_type_col],\n",
    "                                    gene_id = gene, aminoacid_A = aa[0],\n",
    "                                    aminoacid_C = aa[1],\n",
    "                                    aminoacid_G = aa[2],\n",
    "                                    aminoacid_T = aa[3])\n",
    "        \n",
    "        # For genes\n",
    "        if gene in Genes:\n",
    "            # update genes\n",
    "            Genes[gene].extend(row[pos_col])\n",
    "            #print(gene)\n",
    "            #print(Genes[gene])\n",
    "            #Genes[gene].print()\n",
    "\n",
    "        else:\n",
    "            # Define gene\n",
    "            Genes[gene] = Gene(gene_id=gene, contig = row[contig_col],\n",
    "                               start = row[pos_col], end = row[pos_col])\n",
    "            #Genes[gene].print()\n",
    "            #print(Genes[gene])\n",
    "        \n",
    "        if i > nrows:\n",
    "            break \n",
    "info_fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#print(Groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chose sites based on depth in groups to compare\n",
    "Counts = {}\n",
    "with open(indir + '/snps_depth.txt') as depth_fh:\n",
    "    header = depth_fh.readline()\n",
    "    header = header.rstrip()\n",
    "    header = header.split('\\t')\n",
    "    \n",
    "    # Get sample and column indices\n",
    "    samples = header[1:]\n",
    "    indices = {}\n",
    "    for s in samples:\n",
    "        indices[s] = header.index(s)\n",
    "    print(indices)\n",
    "    \n",
    "    \n",
    "    depth_reader = csv.reader(depth_fh, delimiter = '\\t')\n",
    "    i = 0\n",
    "    for row in depth_reader:\n",
    "        i += 1\n",
    "        #print(row)\n",
    "        \n",
    "        site_id = row[0]\n",
    "        #print(site_id)\n",
    "        \n",
    "        # Get all counts\n",
    "        counts = row[1:]\n",
    "        counts = list(map(int,counts))\n",
    "        #print(counts)\n",
    "        \n",
    "        counts = [int(c >= min_count) for c in counts]\n",
    "        \n",
    "        # Get counts per group\n",
    "        samples1 = [int(counts[ indices[l] - 1 ]) for l in Groups[group1]]\n",
    "        samples2 = [int(counts[ indices[l] - 1 ]) for l in Groups[group2]]\n",
    "        samples1 = sum(samples1)\n",
    "        samples2 = sum(samples2)\n",
    "        #print(samples1)\n",
    "        #print(samples2)\n",
    "        if not ((samples1 > 0 and samples2 > 0) and (samples1 > 1 or samples2 > 1)):\n",
    "            # delete\n",
    "            #print(site_id)\n",
    "            if site_id in Sites:\n",
    "                del Sites[site_id]\n",
    "        else:\n",
    "            # NOTE: ASSUMING SAME ORDER IN SAMPLES BETWEEN SITES\n",
    "            Counts[site_id] = counts\n",
    "        \n",
    "        if i > nrows:\n",
    "            break \n",
    "\n",
    "depth_fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read frequencies and calculate \n",
    "print(Groups)\n",
    "MK = {}\n",
    "with open(indir + '/snps_freq.txt') as freqs_fh:\n",
    "    header = freqs_fh.readline()\n",
    "    header = header.rstrip()\n",
    "    header = header.split('\\t')\n",
    "    \n",
    "    # Get sample and column indices\n",
    "    samples = header[1:]\n",
    "    indices = {}\n",
    "    for s in samples:\n",
    "        indices[s] = header.index(s)\n",
    "    print(indices)\n",
    "    print(header)\n",
    "    \n",
    "    freqs_reader = csv.reader(freqs_fh, delimiter = '\\t')\n",
    "    i = 0\n",
    "    for row in freqs_reader:\n",
    "        i += 1\n",
    "        \n",
    "        # Check if site was selected based on sites\n",
    "        site_id = row[0]\n",
    "        if not site_id in Sites:\n",
    "            #print(\"==Skipping\")\n",
    "            continue\n",
    "        \n",
    "        #print(\"==========================\")\n",
    "        gene = Sites[site_id].gene_id\n",
    "        s_type = Sites[site_id].substitution_type()\n",
    "        present_index = np.array(Counts[site_id])\n",
    "        group_index = np.array([Samples[s][0] for s in samples])\n",
    "        #print(row)\n",
    "        #print(site_id)\n",
    "        #allele_frequencies = row.split()\n",
    "        #print(\"Major Allele: {}\".format(Sites[site_id].major_allele))\n",
    "        #print(\"Minor Allele: {}\".format(Sites[site_id].minor_allele))\n",
    "        #print(\"Substitution type: {}\".format(s_type))\n",
    "        #print(\"Gene: {}\".format(gene))\n",
    "        #print(present_index)\n",
    "        #print(group_index)\n",
    "        \n",
    "        # Create MKtest if needed\n",
    "        if gene not in MK:\n",
    "            MK[gene]= MKtest(name=gene)\n",
    "            \n",
    "        # find allele per sample\n",
    "        allele_freqs = np.array([int(float(f) < 0.5) for f in row[1:]])\n",
    "        #print(allele_freqs)\n",
    "        \n",
    "        # Remove non covered positions\n",
    "        ii = np.where(present_index)\n",
    "        group_index = group_index[ii]\n",
    "        allele_freqs = allele_freqs[ii]\n",
    "        #print(group_index)\n",
    "        #print(allele_freqs)\n",
    "        \n",
    "        # Count alleles per group\n",
    "        group1_count = allele_freqs[np.where(group_index == group1)].sum()\n",
    "        group2_count = allele_freqs[np.where(group_index == group2)].sum()\n",
    "        #print(group1_count)\n",
    "        #print(group2_count)\n",
    "        \n",
    "        if group1_count > 0 and group2_count > 0:\n",
    "            fixed = False\n",
    "        elif group1_count > 0 or group2_count > 0:\n",
    "            fixed = True\n",
    "#         else:\n",
    "#             print(\"===============\")\n",
    "#             print(row)\n",
    "#             print(group_index)\n",
    "#             print(allele_freqs)\n",
    "#             raise ValueError(\"At least one of the counts must be non-zero\")\n",
    "        \n",
    "        if s_type is 'synonymous':\n",
    "            if fixed:\n",
    "                MK[gene].update(Ds = 1)\n",
    "            else:\n",
    "                MK[gene].update(Ps = 1)\n",
    "        elif s_type is 'non-synonymous':\n",
    "            if fixed:\n",
    "                MK[gene].update(Dn = 1)\n",
    "            else:\n",
    "                MK[gene].update(Pn = 1)\n",
    "        else:\n",
    "            raise ValueError(\"Invalid substitution type\")\n",
    "            \n",
    "        #print(\"==========================\")\n",
    "        \n",
    "        \n",
    "        if i > nrows:\n",
    "            break\n",
    "freqs_fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(outfile,mode='w') as fh:\n",
    "    for gene,mk in MK.items():\n",
    "        #print(gene)\n",
    "        #print(\"\\t\\tFixed\\tPolymorphic\\n\\tSynonymous\\t{}\\t{}\\n\\tnon-synonymous\\t{}\\t{}\".format(mk.Ds,mk.Ps,mk.Dn,mk.Pn))\n",
    "        #ratio = mk.calculate()\n",
    "        oddratio,pval = mk.test()\n",
    "        res = [gene, str(oddratio), str(pval)]\n",
    "        #print(res)\n",
    "        fh.write(\"\\t\".join(res) + \"\\n\")\n",
    "        #alpha = mk.alpha()\n",
    "        #print(\"MK ratio is: {}\".format(str(ratio)))\n",
    "        #print(\"MK alpha is: {}\".format(str(alpha)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
